### ISSUE TITLE
The results of colpali are not deterministic

### ISSUE THREAD

**Issue by SeanLee97** [illuin-tech/vidore-benchmark#38]
Thank you for the great work!

Recently, I tried to run the benchmark for colpali. And I found that the results are not deterministic, i.e., it has different results for different runs. 

I am trying to fix this issue by adding `model.eval()`, but it doesn't help. Could you help look into this issue?

**Comment by tonywu71** [comment]
Hi @SeanLee97 👋🏼 
Can I ask which version of `colpali-engine` and which Hf model name you are using please?

**Comment by SeanLee97** [comment]
Hi @tonywu71, For library, I used `colpali-engine==0.1.1` and `vidore-benchmark==3.4.3`. For model, I used `vidore/colpali-v1.2`. Should I update the colpali-engine to the latest?

**Comment by tonywu71** [comment]
Yes, that's correct! 👍🏼 The [`v0.2.0`](https://github.com/illuin-tech/colpali/releases/tag/v0.2.0) fixes the issue explained by Manu in https://github.com/illuin-tech/vidore-benchmark/issues/38#issuecomment-2392218396.

Personally, I suggest using:
- `colpali-engine==0.3.1`
- `vidore/colpali-v1.2`
- `vidore-benchmark==0.4.0` (optional).

**Comment by SeanLee97** [comment]
> Yes, that's correct! 👍🏼 The [`v0.2.0`](https://github.com/illuin-tech/colpali/releases/tag/v0.2.0) fixes the issue explained by Manu in [#38 (comment)](https://github.com/illuin-tech/vidore-benchmark/issues/38#issuecomment-2392218396).
> 
> Personally, I suggest using:
> 
> * `colpali-engine==0.3.1`
> * `vidore/colpali-v1.2`
> * `vidore-benchmark==0.4.0` (optional).

hi @tonywu71 @ManuelFay, I used the recommended version, but it is still not deterministic. Here is my evaluation code:

```bash
CUDA_VISIBLE_DEVICES=2 vidore-benchmark evaluate-retriever \
    --model-name vidore/colpali \
    --dataset-name vidore/docvqa_test_subsampled \
    --split test
```

I ran it for two times but got different results:
- 57.05
- 56.72

**Comment by ManuelFay** [comment]
Ok we should look into that ! 
I guess you mean the v1.2 in the CLI though right ?

**Comment by ManuelFay** [comment]
@tonywu71 just checked the code in vidore-benchmark - there is no .eval() after loading the Vision Retriever (or other retrievers). Can you update + verify reproducibility please ?

**Comment by tonywu71** [comment]
Good catch @ManuelFay, I'm on it!

**Comment by tonywu71** [comment]
@SeanLee97 Fixed in https://github.com/illuin-tech/vidore-benchmark/pull/42 and in the new patch release [`v.4.0.1`](https://github.com/illuin-tech/vidore-benchmark/releases/tag/v4.0.1) for `vidore-benchmark`. 💪🏼 

FYI, I've changed the model alias to the default ColPali checkpoint to make it less confusing. Here's the new command you should run:

```bash
vidore-benchmark evaluate-retriever \
    --model-name vidore/colpali-v1.2 \
    --dataset-name vidore/docvqa_test_subsampled \
    --split test
```

**Comment by SeanLee97** [comment]
Hi @tonywu71 @ManuelFay , thank you very much for your update! 

Unfortunately, the issue has not been fixed, the results are still not deterministic. I upgraded `vidore-benchmark` to 4.0.1 and ran it 3 times, the results are as follows:
- 0.56665
- 0.56886
- 0.56804

**I tried to fix the random seed, but it didn't help.**

**To solve this issue, I refactor the evaluation:**

1. First, I converted the vidore benchmark to `beir` format via [this script](https://gist.github.com/SeanLee97/afa94a2a96bb89d3ef83cbb499a883d8).
2. Second, I evaluated it with [this evaluation script](https://gist.github.com/SeanLee97/713d35542ba0a74935f9cdaa20674349).

With our evaluation script, the results are deterministic now.

Here is an evaluation example:

```bash
CUDA_VISIBLE_DEVICES=2 python eval_vidore_beir.py \
    --model_type ColPali \
    --model_name vidore/colpaligemma-3b-pt-448-base \
    --adapter_name vidore/colpali-v1.2 \
    --dataset_name mixedbread-ai/vidore-docvqa_test_subsampled \
    --batch_size 4
```

I would appreciate it if you could help check the new evaluation code.

If it is correct, could you rerun the benchmark and refresh the benchmark with the deterministic evaluation code?

Thank you!

**Comment by tonywu71** [comment]
Thanks for the snippets @SeanLee97. You're absolutely right: the BEIR format makes much more sense for the ViDoRe benchmark as our proposed format doesn't scale well.

I'll be working on integrating this ASAP!

**Comment by ManuelFay** [comment]
Fixed in #60 normally - I get perfectly exact results on all datasets except for tatdqa where I got a 0.00003 difference once in 5 times but small enough I think it's just numerical stability with FA2
